tests/keras/backend/backend_test.py

```diff
@@ -177,6 +177,69 @@ def check_composed_tensor_operations(first_function_name,
     assert_list_pairwise(z_list)
 
 
+def check_rnn_operation(step_function_k,
+                        step_function_np,
+                        inputs_np,
+                        initial_states_np,
+                        mask_np=None,
+                        constants_np=None,
+                        **kwargs):
+    inputs_k = K.variable(inputs_np)
+    initial_states_k = [K.variable(s) for s in initial_states_np]
+    if mask_np is not None:
+        mask_k = K.variable(mask_np)
+    else:
+        mask_k = None
+    if constants_np is not None:
+        constants_k = [K.variable(c) for c in constants_np]
+    else:
+        constants_k = None
+
+    last_output_s = []
+    output_s = []
+    last_states_s = []
+    for unroll in [True, False]:
+        last_output_np, output_np, last_states_np = KNP.rnn(
+            step_function_np,
+            inputs_np,
+            initial_states_np,
+            mask=mask_np,
+            constants=constants_np,
+            unroll=unroll,
+            input_length=inputs_np.shape[1] if unroll else None,
+            **kwargs)
+        last_output_k, output_k, last_states_k = K.rnn(
+            step_function_k,
+            inputs_k,
+            initial_states_k,
+            mask=mask_k,
+            constants=constants_k,
+            unroll=unroll,
+            input_length=inputs_np.shape[1] if unroll else None,
+            **kwargs)
+
+        assert len(last_states_np) == len(initial_states_np)
+        assert len(last_states_k) == len(initial_states_k)
+        last_states_k = [K.eval(s) for s in last_states_k]
+        last_output_k = K.eval(last_output_k)
+        output_k = K.eval(output_k)
+
+        assert_allclose(last_output_k, last_output_np, atol=1e-05)
+        assert_allclose(output_k, output_np, atol=1e-05)
+        for s_k, s_np in zip(last_states_k, last_states_np):
+            assert_allclose(s_k, s_np, atol=1e-05)
+
+        last_output_s.append(last_output_k)
+        output_s.append(output_k)
+        last_states_s.append(last_states_k)
+
+    # compare unroll True/False
```

Comment:
I thought about it last time, and I don't think we need those checks, because you already did everything needed in the numpy implementation. KNP.rnn is garanteed to give the same result for `unroll=True` and `unroll=False` because you don't use the variable `unroll` in the implementation. You already compare against KNP.rnn, so you're good to go. This should simplify this function quite a bit.

Do you see what I mean? Maybe I'm not clear.

Reply:
It's clear what you mean, and all things considered I agree - if we also move out the execution of KNP.rnn and run it once.

This is mainly legacy from how the tests look now (prior to this PR), where this comparison is done (although a bit obfuscated and dependent on the order of `kwargs_list`) even though KNP.rnn is independent or unroll.

These things get unclear with the "reference implementation" approach to testing, connecting back to our discussion [here](https://github.com/keras-team/keras/pull/11499#discussion_r230938541). Are we testing _both_ the `KNP` and `K` implementation in this function? Or only `K`,  taking `KNP` as ground truth? If it is the former then it makes sense to run KNP for both `unroll in [True, False]` and assert equality - we shouldn't assume that the argument is not used, rather make sure to expose if that changes. We basically want the latter, but then it could be a problem that there are no tests for KNP as of my previous argumentation. But enough of that for now - I'll fix 👍

Reply:
I think it's fine to test for KNP. As you said in our previous discussions, we can make certain assumptions regarding the numpy backend when the function is simple (K.expand_dims for example). In this case, I believe it is simple to check that `unroll` is not used in KNP.rnn because
* `unroll` can have no meaning in the numpy implementation of RNNs.
* With a decent text editor with python pluging / IDE, the variable `unroll` is marked as `unused variable` so it's fairly easy to notice an error.

To be clear:
* I believe that the numpy backend should get tested for complex functions
* KNP.rnn should get tested for correctness
* It is safe to assume that KNP.rnn doesn't use `unroll`